lib/layout/force-directed-layout.js

```diff
@@ -0,0 +1,204 @@
+
+function layOutGraph(graph) {
+	mGraph = graph;
+	vertices = graph.get('vertices');
+	var edgeList = graph.get('edges');
+	var runner = true;
+	//var loopCount =0;
+	while(runner){
+		//console.log('loop count is: '+loopCount+'\n');
+		change = false;
+		
+		for(var i=0;i<edgeList.length;i++){
+			springForce(edgeList[i]);
+		}
+		if(change === false){
+			runner = false;
+		}
+		//loopCount++;
+	}
+	//defaulting all ports that do not have a position
+	for(var j=0;j < vertices.length; j++){
+		var inputs = vertices[j].get('inputs');
+		var outputs = vertices[j].get('outputs');
+		
+		for(var h =0; h < inputs.length; h++){
+			
+			if(inputs[h].get('position') === null){
+				inputs[h].set('position', new cardinalPortPosition('E', 50));
+			}
+			
+		}
+		for(h =0;h < outputs.length; h++){
+			if(outputs[h].get('position') === null){
+				outputs[h].set('position', new cardinalPortPosition('W', 50));
+			}
+		}
+	}
+}
+
+function springForce(edge){
+	//vertex to
+	var vt = mGraph.getVertexById(edge.get('to').get('ownerVertexId'));
+	//vertex from
+	var vf = mGraph.getVertexById(edge.get('from').get('ownerVertexId'));
+	//x distance between them
+	var xL = edgeXLength(edge);
+	//y distance between them
+	var yL = edgeYLength(edge);
+	//straight line edge length
+	var length = edgeLength(edge);
+	//checks for both pined
+	if(vt.get('isPinned') ===true && vf.get('isPinned')===true){
+		return;
+	}
+	
+	//hacky solution for length being 0
+	if(length === 0){
+		length = 0.0001;
+	}
+	//repulsion force calculation
+	var force = forceStr * (iLength - length) / length;
+	force = force * Math.pow(stretchForce, (vt.get('inputs').length + vt.get('outputs').length + vf.get('inputs').length + vf.get('outputs').length - 4));
+	
+	//repulsion change in x
+	var cx = force * xL;
+	//repulsion change in y
+	var cy = force * yL;
+	//attraction x
+	var dx;
+	//attraction y
+	var dy;
+	//math for min length
+	if(length < 1){
+		dx = Math.random()*10+1;
+		dy = Math.random()*10+1;
+	} else {
+		dx = xL/(length*length);
+		dy = yL/(length*length);
+	}
+	
+	var len = dx * dx + dy * dy;
+	//ensures length is positive
+	if(len>0){
+		len = Math.sqrt(len);
+		dx = dx / len;
+		dy = dy / len;
+	}
+	//repulsion minus attraction
+	cx = cx - dx;
+	cy = cy - dy;
+	if(vt.get('isPinned') ===false && vf.get('isPinned')===false){
+		vt.set('position', {'x': (vt.get('position').x+cx),'y': (vt.get('position').y+cy)});
+		vf.set('position', {'x':vf.get('position').x-cx, 'y':vf.get('position').y-cy});
+
+		if(cx > 1 || cy > 1){
+			change = true;
+		}
+	}else if(vt.get('isPinned') ===true){
+		vf.set('position', {'x':vf.get('position').x-cx-cx, 'y':vf.get('position').y-cy-cy});
+		
+		if(cx > 1 || cy > 1){
+			change = true;
+		}
+	
+	}else if(vf.get('isPinned')===true){
+		vt.set('position', {'x': (vt.get('position').x+cx),'y': (vt.get('position').y+cy)});
```

Comment:
dude this isn't a competition for how much you can get on one line. Things can be much more readable if you do things one at a time, with sensible variable names. e.g. this could be something like
```
var oldPosition = vertex.get('position');
var newPosition = {
  x: oldPosition.x + cx,
  y: oldPosition.y + cy
};

vertex.set('position': newPosition);
```

I'm having a very hard time following your code.